### PR TITLE
Refactor LoadManager to Scheduler

### PR DESCRIPTION
**Pull request by jja725**
### What changes are proposed in this pull request?

Refactoring LoadManager to Scheduler.
Redefine the job progress metrics since total file count is hard to be accurate when we have retry and verification

### Why are the changes needed?

for other jobs

### Does this PR introduce any user facing changes?
updated API in FileSystem
```
  /**
   * Submit a job.
   *
   * @param jobRequest the job request
   * @return job id if job is submitted, empty if a load of the same path already exists
   */
  Optional<String> submitJob(JobRequest jobRequest);

  /**
   * Stop a job.
   *
   * @param jobDescription the job description
   * @return true if job is stopped, false if cannot find job
   */
  boolean stopJob(String jobDescription);

  /**
   * Get progress of a job.
   *
   * @param jobDescription   the job description
   * @param format  progress report format
   * @param verbose whether to return verbose report
   * @return the job progress
   */
  String getLoadProgress(String jobDescription,
      JobProgressReportFormat format, boolean verbose);
```

### CODE REVIEW

File: core/server/master/src/main/java/alluxio/master/scheduler/Scheduler.java [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+        mRunningTasks.put(job, new HashSet<>());
+      }
+    }
+  }
+
+  /**
+   * Stop scheduler.
+   */
+  public void stop() {
+    if (mRunning) {
```
Comment (jja725) [inline core/server/master/src/main/java/alluxio/master/scheduler/Scheduler.java:118]: @tcrain so one thing I want to verify is that we stop rpc server before calling this scheduler.stop so we can safely cleanup the state

**Comment by jja725** [comment]
Tested locally with transferring leadership, we can still get the job information and run the unfinished jobs.

**Review by beinan — APPROVED** [review]
lgtm

File: core/server/master/src/main/java/alluxio/master/job/JobFactoryProducer.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  private JobFactoryProducer() {} // prevent instantiation
+
+  /**
+   * @param request the job request
+   * @param fsMaster the file system master
+   * @return the job factory
+   */
+  public static JobFactory create(JobRequest request, FileSystemMaster fsMaster) {
+    switch (request.getType()) {
+      case "load":
```
Comment (acgtun) [inline core/server/master/src/main/java/alluxio/master/job/JobFactoryProducer.java:33]: should this be an enum or const variable? 

In LoadJobRequest, it has
`private static final String TYPE = "load";`

Comment (acgtun) [reply core/server/master/src/main/java/alluxio/master/job/JobFactoryProducer.java:33]: In LoadJob, it has 
public static final String TYPE = "load";

Comment (jja725) [reply core/server/master/src/main/java/alluxio/master/job/JobFactoryProducer.java:33]: @acgtun Thanks for the suggestion. I'm still thinking about how to add support for a custom job in the future(like service loader). Would have another PR to support that

**Comment by beinan** [comment]
alluxio-bot, merge this please.

**Comment by jja725** [comment]
alluxio-bot, cherry-pick this to dora please

**Comment by Xenorith** [comment]
alluxio-bot, cherry-pick this to dora please

File: core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java [hunk 3]
```diff
@@ -523,28 +525,27 @@ public void needsSync(AlluxioURI path)
   }
 
   @Override
-  public boolean submitLoad(AlluxioURI path, java.util.OptionalLong bandwidth,
-      boolean usePartialListing, boolean verify) {
+  public Optional<String> submitJob(JobRequest jobRequest) {
```
Comment (acgtun) [inline core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java:528]: After changing the client API from submitLoad to submitJob, will it have backward compatible problem when users use the old client version?

**Pull request merged**